10738 ebay

wgaulke opened the pull request:
Attached my first attempt of the ebay order-import route. Contains two maven projects:

**ebay-api-client**

Generated REST-Client for the fulfilment api. Used openapi-generator with okhttp-gson. Small patch to parse taxAddress objects correctly.

**de-metal-camel-ebay-camelroutes**

With the following highlights:

- GetEbayOrdersRouteBuilder.java - Camel Route for ebay imports
- CreateBPartnerUpsertReqForEbayOrderProcessor  - Creates BPartners for an eBay order
- CreateOrderLineCandidateUpsertReqForEbayOrderProcessor - Created OLC for an eBay order

To test:

- EbayOrderProcessingRouteTest - Creates route, adds mock result from ebay service and runs through whole route


**Notes:**

- contains some minor c/p from shopware api => we should consider refactoring with central util classes for shared components.
- Unit test does not check that much yet. I'll add more eBay API examples and a concept for result checking in the future.


However, from my point of view a good chunk as a first pull. Looking forward to your review :)

metas-ts commented on misc/services/camel/de-metas-camel-externalsystems/ebay/de-metas-camel-ebay-camelroutes/src/main/java/de/metas/camel/ebay/processor/ProcessorHelper.java:
```diff
@@ -0,0 +1,67 @@
+				.pInstanceId(JsonMetasfreshId.of(adPInstanceId))
+				.build();
+
+		exchange.getContext().createProducerTemplate()
```
This doesn't work; see https://camel.apache.org/manual/latest/faq/why-does-camel-use-too-many-threads-with-producertemplate.html
Feel free to check out how they meanwhile fixed this over at the shopware module

metas-ts reviewed (commented):
so far so good. additional remarks:
* please import and apply our formatting rules from `metasfresh/misc/dev-support`
* we need the products. i can provide you some sample code on this
* we need to skip over not-yet "finalized" ebay orders. 
  as a hint: in shopware, we do this, and then communicate from camel to metasfresh, from which timestamp onwards the orders need to be re-checked at shopware the next time. if metasfresh gets OLCandidates it already has, it simply ignores them.

wgaulke commented on misc/services/camel/de-metas-camel-externalsystems/ebay/de-metas-camel-ebay-camelroutes/src/main/java/de/metas/camel/ebay/processor/CreateBPartnerUpsertReqForEbayOrderProcessor.java:
```diff
@@ -0,0 +1,166 @@
+
+			ShippingStep shipTo = order.getFulfillmentStartInstructions().get(0).getShippingStep();
+
+			bpartnerLocation.setAddress1(shipTo.getShipTo().getContactAddress().getAddressLine1());
```
Yes, because these fields are marked as required/guaranteed by the ebay api. However, I agree that this could be a bit more defensive just in case.

wgaulke commented on misc/services/camel/de-metas-camel-externalsystems/ebay/de-metas-camel-ebay-camelroutes/src/main/java/de/metas/camel/ebay/processor/GetEbayOrdersProcessor.java:
```diff
@@ -0,0 +1,159 @@
+			
+		} else {
+
+			ProcessorHelper.logProcessMessage(exchange, "Ebay:Failed to aquire access token!" + Instant.now(), request.getAdPInstanceId().getValue());
```
You are right - no need to continue pipeline execution, changed that.

wgaulke commented:
@metas-ts 

Wanted to give you an update.

* Applied your formatting rules, sorry :)
* Added processor to create products from ebay orderLines
  * Question: area prices nessary too? Or is it ok, if prices are just pushed in by the subsequent OLC upsert?

> we need to skip over not-yet "finalized" ebay orders.
WIP, still open.

metas-ts commented:
hi @wgaulke 

> Question: are prices nessary too? Or is it ok, if prices are just pushed in by the subsequent OLC upsert?

We need product-prices for their `de.metas.common.pricing.v2.productprice.TaxCategory`.
The actual price amount can be zero on the product-price-upsert, if you provide that price (with IsManualPrice) in the OLC upsert.

wgaulke commented:
@metas-ts 

With my last pushes I added:
* Price upsert (current limited to fixed tax-rate)
* JsonRuntimeParameterUpsert based on most recent order TS (like shopware, as you recommended).

Feel free to take a look. Next, I'll improve the integration tests.

wgaulke commented:
@metas-ts another question: is there any preferred json client lib to use? https://openapi-generator.tech/docs/generators/java 

I noticed that the build fails because of some dependency issues. Switching to another client lib could fix that easily.

wgaulke commented:
@metas-ts I reopened this pull request. Changes since last version:

* rebased on master to integrate changes made to external-system setup
* added configuration window for ebay creds
* fix unit test to reflect latest api changes / testing 
* tested complete integration flow

metas-ts reviewed (approved):
good to integrate after successful build 🎉

wgaulke commented on misc/services/camel/de-metas-camel-externalsystems/ebay/de-metas-camel-ebay-camelroutes/README.md:
```diff
@@ -0,0 +1,303 @@
+
+
+
+| Nr. | Field-Name | Occurence E | M | Metas OC-Field | Metas Payment Field | Notes |
```
Good remark, actually no. I'll update it with latest values used in the processor implementations. Values should not change much in the future.

wgaulke commented on misc/services/camel/de-metas-camel-externalsystems/ebay/de-metas-camel-ebay-camelroutes/src/main/java/de/metas/camel/externalsystems/ebay/EbayConstants.java:
```diff
@@ -0,0 +1,76 @@
+	// default values
+	String DATA_SOURCE_INT_EBAY = "int-Ebay";
+	String DEFAULT_DELIVERY_RULE = "A"; // FIXME: magic name
+	String DEFAULT_DELIVERY_VIA_RULE = "S"; // FIXME: magic name
```
For me the values "A" and "S" were kind of magic - once you get to know metasfresh better they start to make sense.

If ok, I'd like to keep the FIXME remark to remember changing it in the future :)

wgaulke commented:
🎉 Thanks for the review and your support. This is huge for us 😊
